File: python/sag_mill_0.py
import logging
import os
import keras
import numpy as np
from keras import layers
from keras.models import Sequential
from keras.optimizers import RMSprop
from matplotlib import pyplot as plt
from pandas import read_csv

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG, format='%(asctime)s %(levelname)s %(message)s')

os.environ['TF_CPP_MIN_LOG_LEVEL']='2'

def main():
	logger.info("Loading dataset...")
	float_data = get_data('sag_mill_0.csv')

	logger.info("Normalizing...")
	data = normalize(float_data)

	lookback = 15
	step = 1
	delay = 10
	batch_size = 128

	train_gen = generator(data, lookback=lookback, delay=delay, min_index=None, max_index=300_000, step=step, batch_size=batch_size)
	val_gen = generator(data, lookback=lookback, delay=delay, min_index=300_001, max_index=400_000, step=step, batch_size=batch_size)
	test_gen = generator(data, lookback=lookback, delay=delay, min_index=400_001, max_index=None, step=step, batch_size=batch_size)

	logger.info("Creating model...")
	model = Sequential()
	model.add(layers.GRU(32, dropout=0.1, recurrent_dropout=0.5, return_sequences=True, input_shape=(None, data.shape[-1])))
	model.add(layers.GRU(64, activation='relu', dropout=0.1, recurrent_dropout=0.5))
	model.add(layers.Dense(2))

	logger.info("Compiling model...")
	model.compile(optimizer=RMSprop(), loss='mae')

	val_steps = (400_000 - 300_001 - lookback)
	history = model.fit_generator(train_gen, steps_per_epoch=500, epochs=50, validation_data=val_gen, validation_steps=val_steps, callbacks=model_callbacks())

	loss = history.history['loss']
	val_loss = history.history['val_loss']
	epochs = range(1, len(loss) + 1)

	plt.figure()
	plt.plot(epochs, loss, 'bo', label='Training loss')
	plt.plot(epochs, val_loss, 'b', label='Validation loss')
	plt.show()

def model_callbacks():
	keras.callbacks.ModelCheckpoint(filepath='sag_mill_0.h5', monitor='val_loss', save_best_only=True,)
	earlystop_callback = keras.callbacks.EarlyStopping(monitor='acc', patience=1)
	plateau_callback = keras.callbacks.ReduceLROnPlateau(monitor='val_loss', factor=0.1, patience=10)
	tensorboard_callback = keras.callbacks.TensorBoard(log_dir='sag_mill_0')
	return [earlystop_callback, plateau_callback, tensorboard_callback]

def get_data(csv_file):
	df = read_csv(csv_file, na_values=[''])
	df.drop(['time'], 1, inplace=True)
	# df = df.replace('', np.NaN)
	df.dropna(inplace=True)
	logger.info("Checking for NaN's...")
	logger.debug(df.isnull().any())
	return df

def normalize(data):
	mean = data.mean(axis=0)
	data -= mean
	std = data.std(axis=0)
	data /= std
	return data

def generator(data, lookback, delay, min_index, max_index, shuffle=False, batch_size=128, step=1):
	if min_index is None:
		min_index = 0
	if max_index is None:
		max_index = len(data) - delay - 1
	index = min_index + lookback
	while True:
		if shuffle:
			rows = np.random.randint(min_index + lookback, max_index, size=batch_size)
		else:
			if index + batch_size >= max_index:
				index = min_index + lookback
			rows = np.arange(index, min(index + batch_size, max_index))
			index += len(rows)

		x = np.zeros((len(rows), lookback//step, data.shape[1]))
		y = np.zeros((len(rows), 2))
		for i, row in enumerate(rows):
			indices = range(rows[i] - lookback, rows[i], step)
			x[i] = data.values[indices]
			y[i,0] = data.values[rows[i] + delay][2] # Mill main drive - Main motor current
			y[i,1] = data.values[rows[i] + delay][5] # Mill weight

		yield x, y

if __name__ == "__main__":
	# execute only if run as a script
	main()

File: Keras-R/sag_mill_0.R
# Targets:
# Mill main drive - Main motor current
# Mill weight

library(readr)
library(ggplot2)
library(magrittr)
library(keras)

setwd("C:/Users/drexa/git/machine-learning/Keras-R")

data_dir <- "./"
filename <- file.path(data_dir, "sag_mill.csv")
csv_data <- read_csv(filename, na="")
dropna_csv_data <- na.omit(csv_data)
na_summary <- sapply(dropna_csv_data, function(x) sum(is.na(x)))
float_data <- data.matrix(dropna_csv_data[,-1])

message("Center and scale...")
data <- scale(float_data, center = TRUE, scale = TRUE)

generator <- function(data, lookback, lookahead, min_index, max_index, shuffle = FALSE, batch_size = 128, step = 1) {
  if (is.null(max_index)) max_index <- nrow(data) - lookahead - 1
  index <- min_index + lookback
  function() {
    if (shuffle) {
      rows <- sample(c((min_index+lookback):max_index), size = batch_size)
    } else {
      if (index + batch_size >= max_index)
        index <<- min_index + lookback
      rows <- c(index:min(index+batch_size, max_index))
      index <<- index + length(rows)
    }
    n_columns <- dim(data)[[2]]
    n_samples <- dim(x)[[2]]
    target_col <- 2
    x <- array(0, dim = c(length(rows), lookback/step, n_columns)) # [batchsize, 240, 14]
    y <- array(0, dim = c(length(rows)))
    for (i in 1:length(rows)) {
      indices <- seq(rows[[i]] - lookback, rows[[i]], length.out = n_samples)
      x[i,,] <- data[indices,]
      y[[i]] <- data[rows[[i]] + lookahead, target_col] # [lookahead row, target]
    }
    list(x, y)
  }
}

lookback <- 15
lookahead <- 10
batch_size <- 128

train_gen <- generator(data, lookback = lookback, lookahead = lookahead, min_index = 1, max_index = 300000, shuffle = FALSE)
val_gen = generator(data, lookback = lookback, lookahead = lookahead, min_index = 300001, max_index = 400000)
test_gen <- generator(data, lookback = lookback, lookahead = lookahead, min_index = 400001, max_index = NULL)

model <- keras_model_sequential() %>%
    layer_gru(units = 32, dropout = 0.1, recurrent_dropout = 0.5, return_sequences = TRUE, input_shape = list(NULL, dim(data)[[2]])) %>%
    layer_gru(units = 64, activation = "relu", dropout = 0.1, recurrent_dropout = 0.5) %>%
    layer_dense(units = 1)

message("Compiling model...")
model %>% compile(optimizer = optimizer_rmsprop(), loss = "mae")

model_callbacks <- function() {
    callback_model_checkpoint <- callback_model_checkpoint("sag_mill_0.h5")
    callback_early_stopping <- callback_early_stopping(monitor="acc", patience=1)
    callback_reduce_lr_on_plateau <- callback_reduce_lr_on_plateau(monitor = "val_loss", factor = 0.1)
    callback_tensorboard <- callback_tensorboard(log_dir="tensorboard_sag_mill_0")
    list(callback_model_checkpoint, callback_early_stopping, callback_reduce_lr_on_plateau, callback_tensorboard)
}

message("Training...")
val_steps <- (400000 - 300001 - lookback)/batch_size
model_callbacks <- model_callbacks()
history <- model %>% fit_generator(train_gen, 
                                   steps_per_epoch = 500, 
                                   epochs = 40,
                                   callbacks = model_callbacks,
                                   validation_data = val_gen, 
                                   validation_steps = val_steps
                                   )

save_model_weights_hdf5(model, "sag_mill_0.h5", overwrite = TRUE)

plot(history)
